util/src/main/java/io/kubernetes/client/informer/cache/Controller.java

```diff
@@ -0,0 +1,137 @@
+
+  /** how we actually process items from the queue */
+  private PopProcessFunc<Deltas<ApiType>> processFunc;
+
+  private ScheduledExecutorService reflectExecutor;
+
+  private ScheduledExecutorService resyncExecutor;
+
+  private ScheduledFuture resyncFuture;
+
+  private Class<ApiType> apiTypeClass;
+
+  private ScheduledFuture reflectorFuture;
+
+  public Controller(
+      Class<ApiType> apiTypeClass,
+      Queue queue,
+      ListerWatcher listerWatcher,
+      PopProcessFunc<Deltas<ApiType>> processFunc,
+      ResyncFunc resyncFunc,
+      long fullResyncPeriod) {
+    this.queue = queue;
+    this.listerWatcher = listerWatcher;
+    this.apiTypeClass = apiTypeClass;
+    this.processFunc = processFunc;
+    this.resyncFunc = resyncFunc;
+    this.fullResyncPeriod = fullResyncPeriod;
+
+    // starts one daemon thread for reflector
+    this.reflectExecutor =
+        Executors.newSingleThreadScheduledExecutor(
+            new ThreadFactoryBuilder()
+                .setNameFormat("controller-reflector-" + apiTypeClass.getName() + "-%d")
+                .build());
+
+    // starts one daemon thread for resync
+    this.resyncExecutor =
+        Executors.newSingleThreadScheduledExecutor(
+            new ThreadFactoryBuilder()
+                .setNameFormat("controller-resync-" + apiTypeClass.getName() + "-%d")
+                .build());
+  }
+
+  public void run() {
+    log.info("informer-Controller#ready to run resync & reflector runnable");
+
+    // start the resync runnable
+    ResyncRunnable resyncRunnable = new ResyncRunnable(queue, resyncFunc);
+
+    if (fullResyncPeriod > 0) {
+      resyncFuture =
+          resyncExecutor.scheduleAtFixedRate(
+              resyncRunnable::run, fullResyncPeriod, fullResyncPeriod, TimeUnit.MILLISECONDS);
+    }
+
+    // TODO(yue9944882): proper naming for reflector
+    reflector = new ReflectorRunnable<ApiType, ApiListType>(apiTypeClass, listerWatcher, queue);
+
+    reflectorFuture =
+        reflectExecutor.scheduleWithFixedDelay(
+            reflector::run, 0L, DEFAULT_PERIOD, TimeUnit.MILLISECONDS);
+
+    // start the process loop
+    this.processLoop();
+  }
+
+  /** stops the resync thread pool firstly, then stop the reflector */
+  public void stop() {
+
+    reflector.stop();
+    reflectorFuture.cancel(true);
+    reflectExecutor.shutdown();
+
+    if (resyncFuture != null) {
+      resyncFuture.cancel(true);
+      resyncExecutor.shutdown();
+    }
+  }
+
+  /** returns true if the queue has been resycned */
+  public boolean hasSynced() {
+    return this.queue.hasSynced();
+  }
+
+  /** returns the latest resource version watched by controller */
+  public String lastSyncResourceVersion() {
+    if (reflector == null) {
+      return "";
+    }
+    return reflector.getLastSyncResourceVersion();
+  }
+
+  /** processLoop drains the work queue. */
+  private void processLoop() {
+    while (true) {
+      try {
+        Deltas deltas = this.queue.pop(this.processFunc);
```

Comment:
You don't actually do anything with `deltas`, remove declaration and inline.

Reply:
make sense. updated